pkg/operator/controllerimagepruner.go

```diff
@@ -0,0 +1,423 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/util/wait"
+	kubeinformers "k8s.io/client-go/informers"
+	kubeset "k8s.io/client-go/kubernetes"
+	appsset "k8s.io/client-go/kubernetes/typed/apps/v1"
+	batchset "k8s.io/client-go/kubernetes/typed/batch/v1beta1"
+	coreset "k8s.io/client-go/kubernetes/typed/core/v1"
+	rbacset "k8s.io/client-go/kubernetes/typed/rbac/v1"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+
+	configapiv1 "github.com/openshift/api/config/v1"
+	imageregistryv1 "github.com/openshift/api/imageregistry/v1"
+	regopset "github.com/openshift/client-go/imageregistry/clientset/versioned"
+	regopinformers "github.com/openshift/client-go/imageregistry/informers/externalversions"
+
+	"github.com/openshift/cluster-image-registry-operator/defaults"
+	regopclient "github.com/openshift/cluster-image-registry-operator/pkg/client"
+	"github.com/openshift/cluster-image-registry-operator/pkg/parameters"
+	"github.com/openshift/cluster-image-registry-operator/pkg/resource"
+	"github.com/openshift/cluster-image-registry-operator/pkg/resource/object"
+	"github.com/openshift/cluster-image-registry-operator/pkg/resource/strategy"
+)
+
+const (
+	imagePrunerWorkQueueKey = "imageprunerchanges"
+)
+
+var (
+	defaultPrunerSuspend                    = false
+	defaultPrunerKeepTagRevisions           = 3
+	defaultPrunerSuccessfulJobsHistoryLimit = int32(3)
+	defaultPrunerFailedJobsHistoryLimit     = int32(3)
+)
+
+// NewImagePrunerController returns a controller for openshift image pruner.
+func NewImagePrunerController(kubeconfig *restclient.Config) (*ImagePrunerController, error) {
+	namespace, err := regopclient.GetWatchNamespace()
+	if err != nil {
+		klog.Fatalf("failed to get watch namespace: %s", err)
+	}
+
+	p := Parameters(namespace)
+
+	listers := &regopclient.ImagePrunerControllerListers{}
+	clients := &regopclient.Clients{}
+	c := &ImagePrunerController{
+		kubeconfig: kubeconfig,
+		params:     *p,
+		generator:  resource.NewImagePrunerGenerator(kubeconfig, clients, listers, p),
+		workqueue:  workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), imagePrunerWorkQueueKey),
+		listers:    listers,
+		clients:    clients,
+	}
+
+	// Initial event to bootstrap the pruner if it doesn't exist.
+	c.workqueue.AddRateLimited(imagePrunerWorkQueueKey)
+
+	return c, nil
+}
+
+// ImagePrunerController keeps track of openshift image pruner components.
+type ImagePrunerController struct {
+	kubeconfig *restclient.Config
+	params     parameters.Globals
+	generator  *resource.ImagePrunerGenerator
+	workqueue  workqueue.RateLimitingInterface
+	listers    *regopclient.ImagePrunerControllerListers
+	clients    *regopclient.Clients
+}
+
+func (c *ImagePrunerController) createOrUpdateResources(cr *imageregistryv1.ImagePruner) error {
+	return c.generator.Apply(cr)
+
```

Comment:
Just curious, why do you add empty lines before `}`?

Reply:
I deleted a bunch of stuff from that function but missed the newline.